### PR TITLE
stm32: adc: fix build of ADC channel setup of STM32L4 series

### LINKED ISSUE #43750: ADC Driver build is broken for STM32L412
**Describe the bug**
Trying to build the ADC driver for a system with an STM32L412 results in an error.

**To Reproduce**
1. Use ADC driver in a project based upon a STM32L412
2. Build the project

**Expected behavior**
ADC3 should not be used.

**Impact**
Breaks the build for projects using the ADC driver and a STM32L412

**Logs and console output**

    zephyr/drivers/adc/adc_stm32.c: In function 'adc_stm32_setup_channels':
    zephyr/drivers/adc/adc_stm32.c:858:29: error: 'ADC3' undeclared (first use in this function); did you mean 'ADC1'?
      858 |  if ((adc == ADC1 || adc == ADC3) &&
          |                             ^~~~
          |                             ADC1

**Environment (please complete the following information):**
Zephyr revision df443f1d29e49f69708da1dfb3018ae8d1ec3f30

### PR DESCRIPTION
**Pull request by benediktibk**
Fixes #43750

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

### CODE REVIEW

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -852,6 +852,29 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
...
 #elif defined(CONFIG_SOC_SERIES_STM32L4X)
+#if defined(CONFIG_SOC_STM32L412XX) || \
+	defined(CONFIG_SOC_STM32L422XX) || \
+	defined(CONFIG_SOC_STM32L432XX) || \
+	defined(CONFIG_SOC_STM32L433XX) || \
+	defined(CONFIG_SOC_STM32L452XX) || \
+	defined(CONFIG_SOC_STM32L462XX) || \
+	defined(CONFIG_SOC_STM32L4S5XX) || \
+	defined(CONFIG_SOC_STM32L4R5XX) || \
+	defined(CONFIG_SOC_STM32L4R9XX)
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:863]: What about `#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)` ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (benediktibk) [reply drivers/adc/adc_stm32.c:863]: I would then go for something like?

    #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)  && DT_NODE_HAS_STATUS(DT_NODELABEL(adc3), okay)
    ...
    #elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
    ...
    #endif

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -855,13 +855,24 @@ static void adc_stm32_setup_channels(const struct device *dev, uint8_t channel_i
...
 	}
+#elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
+	if ((adc == ADC1) &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
+		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_TEMPSENSOR);
+	} else if (adc == ADC1 &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_VREFINT) == channel_id) {
+		adc_stm32_set_common_path(dev, LL_ADC_PATH_INTERNAL_VREFINT);
+	}
+#endif
```
[the remark this answers is not in the document]
Comment (benediktibk) [reply drivers/adc/adc_stm32.c:875]: Me neither. What about a macro function, e.g.

    #if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
	 DT_NODE_HAS_STATUS(DT_NODELABEL(adc3), okay)
    #define IS_ADC_WITH_TEMPERATURE(adc) (adc == ADC1 || adc == ADC3)
    #elif DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay)
    #define IS_ADC_WITH_TEMPERATURE(adc) (adc == ADC1)
    #else
    #define IS_ADC_WITH_TEMPERATURE(adc) (false)
    #endif

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (FRASTM) [reply drivers/adc/adc_stm32.c]: Note that some stm32 mcus like stm32G4 or U5 have TEMP sensor on other ADC than 3: 
```
- The internal reference voltage (V REFINT ), connected to ADC1, 3, 4 and 5
- The internal temperature sensor (V TS ), connected to ADC1 and 5
```

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -798,93 +798,77 @@ static void adc_stm32_setup_speed(const struct device *dev, uint8_t id,
...
-#if defined(CONFIG_SOC_SERIES_STM32G4X)
-	/* In the G4 model, once can access internal temperature on ADCs 1 and 5. */
-	if ((adc == ADC1 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) ||
-		(adc == ADC5 &&
-		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR_ADC5) == channel_id)) {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && \
+	DT_PROP(DT_NODELABEL(adc1), has_temp_channel)
+	if (adc == ADC1 &&
+		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```
Comment (erwango) [inline drivers/adc/adc_stm32.c:807]: From this point, you should be able to do a "mere" (modulo the series api adaptation):
```
	if (config->has_temp_channel &&
		__LL_ADC_CHANNEL_TO_DECIMAL_NB(ADC_CHANNEL_TEMPSENSOR) == channel_id) {
```
This requires adding fields to device config struct, but likely worse the gain  considering this function.

File: dts/arm/st/l4/stm32l4r5.dtsi [hunk 4]
```diff
@@ -312,3 +312,8 @@
 		label = "OTGFS_PHY";
 	};
 };
+
+&adc1 {
+	has-temp-channel;
+	has-vref-channel;
+};
```
Comment (erwango) [inline dts/arm/st/l4/stm32l4r5.dtsi:319]: Please don't add phandle fixups in this file, but rather overwrite the nodes directly:
```
		adc1: adc@foo{
			has-temp-channel;
			has-vref-channel;
		};
	};
};
```

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this nice fix & addition.
I guess new property names could be bikeshed but given the overall improvement, let's go for it.
One request: please  split the dts/bindings from the driver bits into 2 separate commits.
Otherwise LGTM

**Comment by benediktibk** [comment]
> Thanks for this nice fix & addition. I guess new property names could be bikeshed but given the overall improvement, let's go for it. One request: please split the dts/bindings from the driver bits into 2 separate commits. Otherwise LGTM

Thanks for guiding me towards this approach, looks way better now than previously.

**Review by erwango — COMMENTED** [review]
This is nitpicky, but in general we recommend, for bisectibility and ease of the review:
- put the dts and bindings in same commit (some also split them a part) > Having dts and bindings in same commit eases the review, we can see that dts additions match bindings
- put the drivers bits in a dedicated following commit > This way driver is functional before and after the change (... in this case driver was broken but, yet...)

**Review by gmarull — APPROVED** [review]
LGTM. I'll save this as an example of HAL not helping, but increasing code complexity.

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**